src/utils/tcti.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright 2019 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+use crate::response_code::{Error, WrapperErrorKind};
+use std::convert::TryFrom;
+use std::ffi::CString;
+use std::net::IpAddr;
+use std::path::PathBuf;
+
+const DEVICE: &str = "device";
+const MSSIM: &str = "mssim";
+const TABRMD: &str = "tabrmd";
+
+// Possible TCTI to use with the ESYS API.
+/// Placeholder TCTI types that can be used when initialising a `Context` to determine which
+/// interface will be used to communicate with the TPM.
+#[derive(Clone, Debug)]
+pub enum Tcti {
+    Device(DeviceConfig),
+    Mssim(MssimConfig),
+    Tabrmd(TabrmdConfig),
+}
+
+impl TryFrom<Tcti> for CString {
+    type Error = Error;
+
+    fn try_from(tcti: Tcti) -> Result<Self, Error> {
+        let tcti_name = match tcti {
+            Tcti::Device(..) => DEVICE,
+            Tcti::Mssim(..) => MSSIM,
+            Tcti::Tabrmd(..) => TABRMD,
+        };
+
+        let tcti_conf = match tcti {
+            Tcti::Mssim(config) => {
+                if let Some(ServerAddress::Hostname(name)) = &config.host {
+                    if !hostname_validator::is_valid(name) {
+                        return Err(Error::WrapperError(WrapperErrorKind::InvalidParam));
+                    }
+                }
+                format!(
+                    ":host={},port={}",
+                    config.host.unwrap_or_default(),
+                    config.port.unwrap_or(2321)
+                )
+            }
+            Tcti::Device(DeviceConfig { path: Some(path) }) => format!(
+                ":{}",
+                path.to_str()
+                    .ok_or(Error::WrapperError(WrapperErrorKind::InvalidParam))?,
+            ),
+            Tcti::Device(DeviceConfig { path: None }) => String::default(),
+            Tcti::Tabrmd(config) => format!(
+                ":bus_name={},bus_type={}",
+                config
+                    .bus_name
+                    .unwrap_or_else(|| String::from("com.intel.tss2.Tabrmd")),
+                config.bus_type.unwrap_or(BusType::System)
+            ),
+        };
+
+        CString::new(format!("{}{}", tcti_name, tcti_conf))
+            .or(Err(Error::WrapperError(WrapperErrorKind::InvalidParam)))
+    }
+}
+
+/// Configuration for a Device TCTI context
+///
+/// The default configuration lets the library use its default path.
+#[derive(Clone, Debug, Default)]
+pub struct DeviceConfig {
+    /// Path to the device node to connect to
+    ///
+    /// If set to `None`, the default location is used
```

Comment:
It would be nice to list what default locations are used, if you can find it somewhere standard. When browsing the code I think I saw that `/dev/tpmrm0` was tested first, the `/dev/tpm0`

Reply:
I wanted to say that but I had no idea where they are on Windows! Will check